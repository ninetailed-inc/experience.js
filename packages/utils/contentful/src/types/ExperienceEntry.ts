import { SafeParseError, SafeParseSuccess, z } from 'zod';
import { Config, ConfigLike } from '@ninetailed/experience.js-utils';

import { EntrySchema, EntryLike, Entry } from './Entry';
import { AudienceEntry } from './AudienceEntry';
import { EntryFields } from './EntryFields';
import { FieldsOverride } from './FieldsOverride';

export const ExperienceEntryFields = z.object({
  /**
   * The name of the experience (Short Text)
   */
  nt_name: z.string(),

  /**
   * The description of the experience (Short Text)
   */
  nt_description: z.string().optional().nullable(),

  /**
   * The type if the experience (nt_experiment | nt_personalization)
   */
  nt_type: z.union([z.string(), z.string()]),

  /**
   * The config of the experience (JSON)
   */
  nt_config: Config.optional()
    .nullable()
    .default(null)
    .transform((val) => {
      return (
        val ?? {
          traffic: 0,
          distribution: [0.5, 0.5],
          components: [],
          sticky: false,
        }
      );
    }),

  /**
   * The audience of the experience (Audience)
   */
  nt_audience: AudienceEntry.optional().nullable(),

  /**
   * All used variants of the experience (Contentful references to other Content Types)
   */
  nt_variants: z.array(EntrySchema).optional(),
});
export type ExperienceFieldsLike<
  VariantFields extends EntryFields = EntryFields
> = Omit<z.infer<typeof ExperienceEntryFields>, 'nt_variants' | 'nt_config'> & {
  nt_variants?: EntryLike<VariantFields>[];
  nt_config?: ConfigLike;
};
export type ExperienceFields<VariantFields extends EntryFields = EntryFields> =
  Omit<z.infer<typeof ExperienceEntryFields>, 'nt_variants'> & {
    nt_variants: Entry<VariantFields>[];
  };

export const ExperienceEntrySchema = EntrySchema.extend({
  fields: ExperienceEntryFields,
});

type ExperienceEntryInput = z.input<typeof ExperienceEntrySchema>;
export type ExperienceEntryLike<
  VariantFields extends EntryFields = EntryFields
> = FieldsOverride<ExperienceEntryInput, ExperienceFieldsLike<VariantFields>>;

type ExperienceEntryOutput = z.infer<typeof ExperienceEntrySchema>;
export type ExperienceEntry<VariantFields extends EntryFields = EntryFields> =
  FieldsOverride<ExperienceEntryOutput, ExperienceFields<VariantFields>>;

const parse = <T extends EntryFields>(
  input: ExperienceEntryLike<T>
): ExperienceEntry<T> => {
  const output = ExperienceEntrySchema.parse(input);

  return {
    ...output,
    fields: { ...output.fields, nt_variants: input.fields.nt_variants || [] },
  };
};

const safeParse = <T extends EntryFields>(
  input: ExperienceEntryLike<T>
):
  | SafeParseError<ExperienceEntryLike<T>>
  | SafeParseSuccess<ExperienceEntry<T>> => {
  const output = ExperienceEntrySchema.safeParse(input);

  if (!output.success) {
    return output;
  }

  return {
    ...output,
    data: {
      ...output.data,
      fields: {
        ...output.data.fields,
        nt_variants: input.fields.nt_variants || [],
      },
    },
  };
};

export const ExperienceEntry = {
  ...ExperienceEntrySchema,
  parse,
  safeParse,
};
