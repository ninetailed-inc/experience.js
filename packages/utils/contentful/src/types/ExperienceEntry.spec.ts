import { ExperienceEntry, ExperienceEntryLike } from './ExperienceEntry';
import { experienceEntryWithoutVariants } from './fixtures/experienceEntryWithoutVariants';
import { experienceEntryWithoutLinkType } from './fixtures/experienceEntryWithoutLinkType';

describe('ExperienceEntry', () => {
  it('should parse an ExperienceEntry without variants', () => {
    expect(
      ExperienceEntry.parse(
        experienceEntryWithoutVariants as unknown as ExperienceEntryLike
      ).fields.nt_variants
    ).toEqual([]);
  });

  it('should parse an ExperienceEntry without linkType', () => {
    experienceEntryWithoutLinkType.map((entry) => {
      expect(
        ExperienceEntry.parse(entry as unknown as ExperienceEntryLike)
      ).toMatchSnapshot();
    });
  });
});
