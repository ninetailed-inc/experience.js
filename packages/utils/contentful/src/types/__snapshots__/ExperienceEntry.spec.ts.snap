// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperienceEntry should parse an ExperienceEntry without linkType 1`] = `
Object {
  "fields": Object {
    "nt_audience": null,
    "nt_config": Object {
      "components": Array [
        Object {
          "baseline": Object {
            "hidden": false,
            "id": "50LFM6YHR9NoqPilYKEGeT",
          },
          "variants": Array [
            Object {
              "hidden": false,
              "id": "2Q5ZkLPpEs5ZOmEttJMh42",
            },
          ],
        },
      ],
      "distribution": Array [
        0.5,
        0.5,
      ],
      "sticky": false,
      "traffic": 1,
    },
    "nt_description": null,
    "nt_name": "OS Testing",
    "nt_type": "nt_experiment",
    "nt_variants": Array [
      Object {
        "fields": Object {
          "anchor": "professional-professional-art-pass",
          "animation": null,
          "body": Object {
            "content": Array [
              Object {
                "content": Array [
                  Object {
                    "content": Array [],
                    "marks": Array [],
                    "nodeType": "text",
                    "value": "A Professional Art Pass lets you enjoy free entry to hundreds of museums, galleries and historic houses across the UK as well as 50% off major exhibitions.",
                  },
                ],
                "data": Array [],
                "nodeType": "paragraph",
              },
            ],
            "nodeType": "document",
          },
          "buttonLabel": "Get your pass",
          "buttonUrl": "#professional-get-your-pass",
          "campaignOptions": null,
          "category": null,
          "component": "Professional Art Pass",
          "homepageCountdown": null,
          "homepageLayout": "Full Screen",
          "homepagePanelColour": "Art Fund Pastel Pink",
          "homepagePanelOptions": "None",
          "homepageStyle": "Images",
          "images": Array [
            Object {
              "alt": null,
              "bynder": Object {
                "expires": null,
                "id": "D7C80D4A-EA34-4958-A5E98E720AAEC1CB",
              },
              "caption": null,
              "copyright": null,
              "primary": false,
              "srcset": Object {
                "lg": "https://dnztw2jgbuerz.cloudfront.net/D7C80D4A-EA34-4958-A5E98E720AAEC1CB/e72d4ee9468dbde96be601c11d0d6f10-lg.jpg",
                "md": "https://dnztw2jgbuerz.cloudfront.net/D7C80D4A-EA34-4958-A5E98E720AAEC1CB/e72d4ee9468dbde96be601c11d0d6f10-md.jpg",
                "xl": "https://dnztw2jgbuerz.cloudfront.net/D7C80D4A-EA34-4958-A5E98E720AAEC1CB/e72d4ee9468dbde96be601c11d0d6f10-xl.jpg",
                "xs": "https://dnztw2jgbuerz.cloudfront.net/D7C80D4A-EA34-4958-A5E98E720AAEC1CB/e72d4ee9468dbde96be601c11d0d6f10-xs.jpg",
                "xxs": "https://dnztw2jgbuerz.cloudfront.net/D7C80D4A-EA34-4958-A5E98E720AAEC1CB/e72d4ee9468dbde96be601c11d0d6f10-xxs.jpg",
              },
            },
            Object {
              "alt": null,
              "bynder": Object {
                "expires": null,
                "id": "F948EE60-5012-4C0A-B9355CF25E631BD8",
              },
              "caption": null,
              "copyright": null,
              "primary": true,
              "srcset": Object {
                "lg": "https://dnztw2jgbuerz.cloudfront.net/F948EE60-5012-4C0A-B9355CF25E631BD8/62e7be1dda70d6be7f5449b6d0a54158-lg.jpg",
                "md": "https://dnztw2jgbuerz.cloudfront.net/F948EE60-5012-4C0A-B9355CF25E631BD8/62e7be1dda70d6be7f5449b6d0a54158-md.jpg",
                "xl": "https://dnztw2jgbuerz.cloudfront.net/F948EE60-5012-4C0A-B9355CF25E631BD8/62e7be1dda70d6be7f5449b6d0a54158-xl.jpg",
                "xs": "https://dnztw2jgbuerz.cloudfront.net/F948EE60-5012-4C0A-B9355CF25E631BD8/62e7be1dda70d6be7f5449b6d0a54158-xs.jpg",
                "xxs": "https://dnztw2jgbuerz.cloudfront.net/F948EE60-5012-4C0A-B9355CF25E631BD8/62e7be1dda70d6be7f5449b6d0a54158-xxs.jpg",
              },
            },
          ],
          "internalTitle": "Professional - Professional Art Pass",
          "layout": "Image on Left",
          "loginOptions": false,
          "nt_experiences": Array [],
          "title": "Get your pass",
          "video": null,
          "videoId": null,
          "videoType": null,
        },
        "metadata": Object {
          "tagNames": Array [],
          "tags": Array [],
        },
        "sys": Object {
          "contentType": Object {
            "description": "",
            "displayField": "internalTitle",
            "fields": Array [
              Object {
                "id": "internalTitle",
                "localized": false,
                "name": "Internal title",
                "required": true,
                "type": "Symbol",
              },
              Object {
                "id": "component",
                "localized": false,
                "name": "Type",
                "required": true,
                "type": "Symbol",
              },
              Object {
                "id": "images",
                "localized": false,
                "name": "Images",
                "required": false,
                "type": "Object",
              },
              Object {
                "id": "videoId",
                "linkType": "Entry",
                "localized": false,
                "name": "Video",
                "required": false,
                "type": "Link",
              },
              Object {
                "id": "category",
                "localized": false,
                "name": "Pre-title",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "title",
                "localized": false,
                "name": "Title",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "body",
                "localized": false,
                "name": "Summary",
                "required": false,
                "type": "RichText",
              },
              Object {
                "id": "layout",
                "localized": false,
                "name": "Layout",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "homepageLayout",
                "localized": false,
                "name": "Homepage Layout",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "homepageStyle",
                "localized": false,
                "name": "Homepage Style",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "videoType",
                "localized": false,
                "name": "Video Type",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "video",
                "localized": false,
                "name": "Video",
                "required": false,
                "type": "Object",
              },
              Object {
                "id": "animation",
                "localized": false,
                "name": "Animation",
                "required": false,
                "type": "Object",
              },
              Object {
                "id": "homepagePanelColour",
                "localized": false,
                "name": "Homepage Panel Colour",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "homepagePanelOptions",
                "localized": false,
                "name": "Homepage Panel Options",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "homepageCountdown",
                "localized": false,
                "name": "Countdown End Date/Time",
                "required": false,
                "type": "Date",
              },
              Object {
                "id": "buttonLabel",
                "localized": false,
                "name": "Button Label",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "buttonUrl",
                "localized": false,
                "name": "Button URL",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "loginOptions",
                "localized": false,
                "name": "Login options?",
                "required": false,
                "type": "Boolean",
              },
              Object {
                "id": "campaignOptions",
                "localized": false,
                "name": "Campaign Options",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "anchor",
                "localized": false,
                "name": "Anchor Key",
                "required": false,
                "type": "Symbol",
              },
              Object {
                "id": "nt_experiences",
                "items": Object {
                  "linkType": "Entry",
                  "type": "Link",
                },
                "localized": false,
                "name": "Ninetailed",
                "required": false,
                "type": "Array",
              },
            ],
            "name": "Module : Hero Panel",
            "sys": Object {
              "createdAt": "2022-10-17T10:01:55.847Z",
              "environment": Object {
                "sys": Object {
                  "id": "QA",
                  "linkType": "Environment",
                  "type": "Link",
                },
              },
              "id": "moduleHero",
              "revision": 68,
              "space": Object {
                "sys": Object {
                  "id": "idkl71bhxp9d",
                  "linkType": "Space",
                  "type": "Link",
                },
              },
              "type": "ContentType",
              "updatedAt": "2023-08-10T14:54:11.067Z",
            },
          },
          "firstPublishedAt": "2023-04-18T15:57:59.920000Z",
          "id": "2Q5ZkLPpEs5ZOmEttJMh42",
          "publishedAt": "2023-04-19T14:33:35.274000Z",
        },
      },
    ],
  },
  "metadata": Object {
    "tags": Array [],
  },
  "sys": Object {
    "contentType": Object {
      "sys": Object {
        "id": "nt_experience",
        "type": "ContentType",
      },
    },
    "id": "4l5B7sToaNGT0pL8wNAtsP",
  },
}
`;
